go/consensus/tendermint/apps/staking/transactions_test.go

```diff
@@ -104,4 +106,352 @@ func TestReservedAddresses(t *testing.T) {
+
+	err = app.withdraw(ctx, stakeState, &staking.Withdraw{})
+	require.EqualError(err, "staking: forbidden by policy", "allow for reserved address should error")
+}
+
+func TestAllow(t *testing.T) {
+	require := require.New(t)
+	var err error
+
+	now := time.Unix(1580461674, 0)
+	appState := abciAPI.NewMockApplicationState(&abciAPI.MockApplicationStateConfig{})
+	ctx := appState.NewContext(abciAPI.ContextDeliverTx, now)
+	defer ctx.Close()
+
+	stakeState := stakingState.NewMutableState(ctx.State())
+
+	app := &stakingApplication{
+		state: appState,
+	}
+
+	pk1 := signature.NewPublicKey("aaafffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff")
+	addr1 := staking.NewAddress(pk1)
+	pk2 := signature.NewPublicKey("bbbfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff")
+	addr2 := staking.NewAddress(pk2)
+	pk3 := signature.NewPublicKey("cccfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff")
+	addr3 := staking.NewAddress(pk3)
+
+	reservedPK := signature.NewPublicKey("badaffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff")
+	reservedAddr := staking.NewReservedAddress(reservedPK)
+
+	for _, tc := range []struct {
+		msg      string
+		params   *staking.ConsensusParameters
+		txSigner signature.PublicKey
+		allow    *staking.Allow
+		err      error
+	}{
+		{
+			"should fail with disabled transfers",
+			&staking.ConsensusParameters{
+				DisableTransfers: true,
+				MaxAllowances:    42,
+			},
+			pk1,
+			&staking.Allow{
+				Beneficiary:  addr2,
+				AmountChange: *quantity.NewFromUint64(10),
+			},
+			staking.ErrForbidden,
+		},
+		{
+			"should fail with zero max allowances",
+			&staking.ConsensusParameters{
+				MaxAllowances: 0,
+			},
+			pk1,
+			&staking.Allow{
+				Beneficiary:  addr2,
+				AmountChange: *quantity.NewFromUint64(10),
+			},
+			staking.ErrForbidden,
+		},
+		{
+			"should fail with equal addresses",
+			&staking.ConsensusParameters{
+				MaxAllowances: 1,
+			},
+			pk1,
+			&staking.Allow{
+				Beneficiary:  addr1,
+				AmountChange: *quantity.NewFromUint64(10),
+			},
+			staking.ErrInvalidArgument,
+		},
+		{
+			"should fail with reserved signer address",
+			&staking.ConsensusParameters{
+				MaxAllowances: 1,
+			},
+			reservedPK,
+			&staking.Allow{
+				Beneficiary:  addr2,
+				AmountChange: *quantity.NewFromUint64(10),
+			},
+			staking.ErrForbidden,
+		},
+		{
+			"should fail with reserved beneficiary address",
+			&staking.ConsensusParameters{
+				MaxAllowances: 1,
+			},
+			pk1,
+			&staking.Allow{
+				Beneficiary:  reservedAddr,
+				AmountChange: *quantity.NewFromUint64(10),
+			},
+			staking.ErrForbidden,
+		},
+		{
+			"should succeed",
+			&staking.ConsensusParameters{
+				MaxAllowances: 1,
+			},
+			pk1,
+			&staking.Allow{
+				Beneficiary:  addr2,
+				AmountChange: *quantity.NewFromUint64(10),
+			},
+			nil,
+		},
+		{
+			"should succeed (updating existing allowance)",
```

Comment:
I know it's tested elsewhere, but maybe a unit test case with negative update would be nice to have here as well.

Reply:
Added.